src/local_cache.cc

```diff
@@ -271,30 +315,62 @@ LocalCache::Insert(const std::string& key, CacheEntry& entry)
     }
   }
 
-  // Exit early if key already exists to avoid setting up entry unnecessarily
+  return nullptr;  // success
+}
+
+TRITONSERVER_Error*
+LocalCache::Insert(
+    const std::string& key, TRITONCACHE_CacheEntry* entry,
+    TRITONCACHE_Allocator* allocator)
+{
+  std::unique_lock lk(cache_mu_);
+  // total_insertion_latency must be protected by mutex
+  helpers::ScopedTimer timer(total_insertion_latency_us_);
+
+  // Exit early to avoid cost of entry setup
   if (cache_.find(key) != cache_.end()) {
     return TRITONSERVER_ErrorNew(
         TRITONSERVER_ERROR_ALREADY_EXISTS,
-        std::string("Insert failed, key [" + key + "] already exists in cache")
-            .c_str());
+        (std::string("key '") + key + std::string("' already exists")).c_str());
   }
 
-  // Allocate and copy into a chunk from managed buffer for each item in entry
-  // NOTE: probably a cleaner way to do this
-  for (auto& item : entry.items_) {
-    for (auto& [base, byte_size] : item.buffers_) {
-      // Copy triton contents into cache representation for cache to own
-      void* new_base = nullptr;
-      // Request block of memory from cache
-      RETURN_IF_ERROR(Allocate(byte_size, &new_base));
-      // Copy contents into cache-allocated buffer
+  // Parse and validate Triton entry
+  std::vector<Metadata> metadata;
+  RETURN_IF_ERROR(ParseTritonEntry(entry, metadata));
+
+  // Form cache representation of entry from Triton entry
+  bool callback = false;
+  auto lentry = std::make_unique<CacheEntry>();
+  for (size_t idx = 0; idx < metadata.size(); idx++) {
+    const auto& [base, byte_size, attrs] = metadata[idx];
+    // Copy triton contents into cache representation for cache to own
+    void* new_base = nullptr;
+    // Request block of memory from cache
+    RETURN_IF_ERROR(Allocate(byte_size, &new_base));
+    // NOTE: For now, buffers in an entry are expected to either uniformly
+    // all be null, or all be non-null. A mix of null and non-null buffers
+    // may lead to unexpected behavior or error.
+    if (base) {
+      // If buffer is provided, copy directly into cache buffer from it
       std::memcpy(new_base, base, byte_size);
-      // Replace triton pointer with cache-allocated pointer in cache entry
-      base = new_base;
+    } else {
```

Comment:
@GuanLuo from our discussion: Remove this divergence, just add a CacheToBytesAllocator that does this simple copy. This way, allocator is required and callback will always be made.

Reply:
done